timeseries/src/autogluon/timeseries/models/ensemble/greedy_ensemble.py

```diff
@@ -38,31 +38,53 @@ def __init__(
-        data_past = labels.slice_by_timestep(None, -self.metric.prediction_length)
-        data_future = labels.slice_by_timestep(-self.metric.prediction_length, None)
-        self.metric.save_past_metrics(data_past)
+        # Stack predictions for each model into a 3d tensor of shape [num_val_windows, num_rows, num_cols]
+        stacked_predictions = [np.stack(preds) for preds in predictions]
+
+        self.dummy_pred_per_window = []
+        self.evaluator_per_window = []
+        self.data_future_per_window = []
+
+        for window_idx, data in enumerate(labels):
+            dummy_pred = copy.deepcopy(predictions[0][window_idx])
+            # This should never happen; sanity check to make sure that all predictions have the same index
+            assert all(dummy_pred.index.equals(pred[window_idx].index) for pred in predictions)
+
+            self.dummy_pred_per_window.append(dummy_pred)
+
+            evaluator = copy.deepcopy(self.metric)
+            # Split the observed time series once to avoid repeated computations inside the evaluator
+            data_past = data.slice_by_timestep(None, -self.metric.prediction_length)
+            data_future = data.slice_by_timestep(-self.metric.prediction_length, None)
+            evaluator.save_past_metrics(data_past)
+            self.evaluator_per_window.append(evaluator)
+            self.data_future_per_window.append(data_future)
+
         super()._fit(
-            predictions=[d.values for d in predictions],
+            predictions=stacked_predictions,
             labels=data_future,
             time_limit=time_limit,
         )
-        self.dummy_pred = None
-
-    def _calculate_regret(self, y_true, y_pred_proba, metric, dummy_pred=None, sample_weight=None):  # noqa
-        dummy_pred = copy.deepcopy(self.dummy_pred if dummy_pred is None else dummy_pred)
-        dummy_pred[list(dummy_pred.columns)] = y_pred_proba
-        score = metric.score_with_saved_past_metrics(y_true, dummy_pred) * metric.coefficient
+        self.dummy_pred_per_window = None
+        self.evaluator_per_window = None
+        self.data_future_per_window = None
+
+    def _calculate_regret(self, y_true, y_pred_proba, metric=None, sample_weight=None):  # noqa
+        # Compute average score across all validation windows
+        total_score = 0.0
+        for window_idx, data_future in enumerate(self.data_future_per_window):
+            dummy_pred = self.dummy_pred_per_window[window_idx]
+            dummy_pred[list(dummy_pred.columns)] = y_pred_proba[window_idx]
+            evaluator = self.evaluator_per_window[window_idx]
+            total_score += evaluator.score_with_saved_past_metrics(data_future, dummy_pred) * metric.coefficient
+        avg_score = total_score / len(self.data_future_per_window)
         # score: higher is better, regret: lower is better, so we flip the sign
-        return -score
+        return -avg_score
```

Comment:
This change is necessary to fix a bug in computation of metrics like `WQL` and `WAPE`.

Here is an example that demonstrates this problem for `prediction_length=1`, `num_val_windows=1`, `eval_metric="WAPE"`.

| Fold | Actual value | Predicted value |
|--------|--------|--------|
| 0 | 100 | 101 |
| 1 | 1 | 10 |

Old code: 
  1. add errors across all folds
  2. then divide by the total target value across all folds
```
(abs(100 - 101) + abs(1 - 10)) / (abs(1) + abs(100)) = 0.099
```

New code:
 1. divide error for each fold by the total target value for this fold
 2. average normalized errors across folds
```
abs(100 - 101) / abs(100) + abs(1 - 10) / abs(1) = 9.01
```

----
Real-world example

Old code:
```
              model  score_test  score_val  
0  WeightedEnsemble   -0.392193  -0.522395 <- incorrect WAPE value because of bug
1     SeasonalNaive   -0.392193  -0.459639 
2             Naive   -0.392193  -0.459639
```

New code:
```
              model  score_test  score_val  
0  WeightedEnsemble   -0.392193  -0.459639
1     SeasonalNaive   -0.392193  -0.459639 
2             Naive   -0.392193  -0.459639 
```
Because of this bug, we may not select the WeightedEnsemble as the best model for prediction.